File: backend/app/schemas/interaction.py
"""
客戶互動記錄相關的 Pydantic Schemas
"""
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import datetime
from enum import Enum


# ================================
# Enums
# ================================

class InteractionTypeEnum(str, Enum):
    """互動類型"""
    DOCUMENT = "document"
    AUDIO = "audio"
    STATUS_CHANGE = "status_change"


# ================================
# Base Schema
# ================================

class InteractionBase(BaseModel):
    """互動記錄基礎 Schema"""
    customer_id: str = Field(..., description="客戶 ID")
    interaction_type: InteractionTypeEnum = Field(..., description="互動類型")
    title: Optional[str] = Field(None, max_length=255, description="標題")
    notes: Optional[str] = Field(None, description="備註")


# ================================
# Create Schema
# ================================

class InteractionCreate(InteractionBase):
    """建立互動記錄 Schema（非檔案上傳）"""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "customer_id": "customer-123",
                "interaction_type": "status_change",
                "title": "狀態變更：已完成一訪",
                "notes": "客戶對產品表示興趣，已安排二訪"
            }
        }
    )


# ================================
# Response Schemas
# ================================

class InteractionResponse(BaseModel):
    """互動記錄回應 Schema"""
    id: str = Field(..., description="互動記錄 ID")
    customer_id: str = Field(..., description="客戶 ID")
    interaction_type: str = Field(..., description="互動類型")
    title: Optional[str] = Field(None, description="標題")
    notes: Optional[str] = Field(None, description="備註")
    file_path: Optional[str] = Field(None, description="檔案路徑")
    file_name: Optional[str] = Field(None, description="檔案名稱")
    file_size: Optional[int] = Field(None, description="檔案大小（bytes）")
    file_type: Optional[str] = Field(None, description="檔案類型")
    audio_duration: Optional[int] = Field(None, description="音訊時長（秒）")
    transcript_text: Optional[str] = Field(None, description="文字稿")
    created_at: datetime = Field(..., description="建立時間")
    updated_at: Optional[datetime] = Field(None, description="更新時間")
    created_by: Optional[str] = Field(None, description="建立者 ID")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": "interaction-123",
                "customer_id": "customer-123",
                "interaction_type": "document",
                "title": "一訪會議記錄",
                "notes": "討論物件管理需求",
                "file_path": "/storage/interactions/documents/doc_123.pdf",
                "file_name": "meeting_notes.pdf",
                "file_size": 1024000,
                "file_type": "application/pdf",
                "created_at": "2026-03-27T10:00:00Z"
            }
        }
    )


class InteractionUploadResponse(BaseModel):
    """檔案上傳後的互動記錄回應"""
    id: str = Field(..., description="互動記錄 ID")
    customer_id: str = Field(..., description="客戶 ID")
    interaction_type: str = Field(..., description="互動類型")
    title: Optional[str] = Field(None, description="標題")
    file_path: Optional[str] = Field(None, description="檔案路徑")
    file_name: str = Field(..., description="檔案名稱")
    file_size: int = Field(..., description="檔案大小（bytes）", ge=0)
    audio_duration: Optional[int] = Field(None, description="音訊時長（秒）", ge=0)
    notes: Optional[str] = Field(None, description="備註")
    created_at: datetime = Field(..., description="建立時間")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": "interaction-123",
                "customer_id": "customer-123",
                "interaction_type": "audio",
                "title": "客戶訪談錄音",
                "file_path": "interactions/audios/recording_123.mp3",
                "file_name": "recording_20260327.mp3",
                "file_size": 5242880,
                "audio_duration": 300,
                "notes": "討論物件管理需求",
                "created_at": "2026-03-27T10:00:00Z"
            }
        }
    )


class InteractionListResponse(BaseModel):
    """互動記錄列表回應"""
    interactions: List[InteractionResponse] = Field(..., description="互動記錄列表")
    total: int = Field(..., description="總筆數", ge=0)
    page: int = Field(..., description="當前頁碼", ge=1)
    limit: int = Field(..., description="每頁筆數", ge=1)
    total_pages: int = Field(..., description="總頁數", ge=0)

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "interactions": [],
                "total": 10,
                "page": 1,
                "limit": 20,
                "total_pages": 1
            }
        }
    )

File: frontend/src/api/interaction.ts
/**
 * 客戶互動記錄 API
 */
import http from './http'

// 互動類型枚舉
export enum InteractionType {
  DOCUMENT = 'document',
  AUDIO = 'audio',
  STATUS_CHANGE = 'status_change'
}

// 互動記錄介面
export interface Interaction {
  id: string
  customer_id: string
  interaction_type: InteractionType
  title?: string
  notes?: string
  file_path?: string
  file_name?: string
  file_size?: number
  file_type?: string
  audio_duration?: number
  transcript_text?: string
  created_at: string
  updated_at?: string
  created_by?: string
}

// 建立互動記錄請求
export interface InteractionCreate {
  customer_id: string
  interaction_type: InteractionType
  title?: string
  notes?: string
}

// 檔案上傳回應
export interface InteractionUploadResponse {
  id: string
  customer_id: string
  interaction_type: InteractionType
  title?: string
  file_path?: string
  file_name: string
  file_size: number
  audio_duration?: number
  notes?: string
  created_at: string
}

// 列表查詢回應
export interface InteractionListResponse {
  interactions: Interaction[]
  total: number
  page: number
  limit: number
  total_pages: number
}

// 查詢參數
export interface InteractionQueryParams {
  customer_id?: string
  interaction_type?: InteractionType
  page?: number
  limit?: number
}

// 上傳進度回調
export interface UploadProgressCallback {
  (progressEvent: { loaded: number; total?: number; percentage?: number }): void
}

/**
 * 建立互動記錄（非檔案上傳）
 */
export async function createInteraction(data: InteractionCreate): Promise<Interaction> {
  const response = await http.post<Interaction>('/api/v1/interactions', data)
  return response.data
}

/**
 * 上傳檔案並建立互動記錄
 * @param file - 檔案物件
 * @param customerId - 客戶 ID
 * @param interactionType - 互動類型
 * @param title - 標題（可選）
 * @param notes - 備註（可選）
 * @param onProgress - 上傳進度回調（可選）
 */
export async function uploadInteractionFile(
  file: File,
  customerId: string,
  interactionType: InteractionType,
  title?: string,
  notes?: string,
  onProgress?: UploadProgressCallback
): Promise<InteractionUploadResponse> {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('customer_id', customerId)
  formData.append('interaction_type', interactionType)
  if (title) {
    formData.append('title', title)
  }
  if (notes) {
    formData.append('notes', notes)
  }

  const response = await http.post<InteractionUploadResponse>(
    '/api/v1/interactions/upload',
    formData,
    {
      headers: {
        'Content-Type': 'multipart/form-data'
      },
      onUploadProgress: (progressEvent) => {
        if (onProgress && progressEvent.total) {
          const percentage = Math.round((progressEvent.loaded * 100) / progressEvent.total)
          onProgress({
            loaded: progressEvent.loaded,
            total: progressEvent.total,
            percentage
          })
        }
      },
      timeout: 120000 // 120 秒超時（檔案上傳可能較慢）
    }
  )

  return response.data
}

/**
 * 查詢互動記錄列表
 * @param params - 查詢參數
 */
export async function getInteractions(
  params?: InteractionQueryParams
): Promise<InteractionListResponse> {
  const response = await http.get<InteractionListResponse>('/api/v1/interactions', {
    params
  })
  return response.data
}

/**
 * 查詢單一互動記錄
 * @param id - 互動記錄 ID
 */
export async function getInteraction(id: string): Promise<Interaction> {
  const response = await http.get<Interaction>(`/api/v1/interactions/${id}`)
  return response.data
}

/**
 * 刪除互動記錄
 * @param id - 互動記錄 ID
 */
export async function deleteInteraction(id: string): Promise<void> {
  await http.delete(`/api/v1/interactions/${id}`)
}
